rfcs/0180-broken-package-removal.md

```diff
@@ -0,0 +1,94 @@
+---
+feature: broken-package-removal
+start-date: 2024-07-14
+author: Jörg "Mic92" Thalheim
+co-authors:
+shepherd-team: @emilazy @preisi @jopejoe1
+shepherd-leader: @jopejoe1
+related-issues:
+---
+
+# Summary
+
+[summary]: #summary
+
+This RFC defines conditions under which we remove broken packages or unmaintained leaf packages.
+The RFC does not aim to define all possible reasons for removing a package but instead focuses on simple, automatable rules.
+Since this document serves as a guideline for automatic package removal, it also allows for the possibility of removing a package earlier for other reasons.
+
+# Motivation
+
+[motivation]: #motivation
+
+Broken and unmaintained packages still consume valuable time.
+Broken packages need to be evaluated to some extent, and unmaintained packages might still need to be rebuilt when one of their dependencies changes.
+Removing unmaintained or unused code should improve overall package quality and help save resources.
+Since adding packages is easy, removing them should also be simple, ideally automated.
+
+# Detailed Design
+
+[design]: #detailed-design
+
+All broken packages or unmaintained leaf packages will be removed after a full NixOS release cycle.
+For example, if `hello-unmaintained` is marked as unmaintained in 23.11, it can be removed from master after the 24.05 release.
+
+## Broken Packages
+
+Packages that are unconditionally marked as broken on all platforms are candidates for removal.
+If a package has dependent packages, those dependent packages will also be marked as indirectly broken.
+If the dependent packages cannot function without the broken package, they should also be removed.
+For example, if a library is broken and an application depends on it, the application will also be removed within the same time frame.
+
+## Unmaintained Packages
+
+We will add an "unmaintained" warning for all packages with an empty maintainer field that do not have any dependent packages.
```

Comment:
How does this mechanism work exactly? How will this work with packages that have broken metadata/no metadata?
I noticed that ofborg doesn't ping maintainers in cases where modifications aren't done in the actual derivation. I.e. if someone updates the source attributes in https://github.com/NixOS/nixpkgs/blob/247cebfe0bb628849a8781be40f193bf622b877b/pkgs/development/tools/pnpm/default.nix, I won't get pinged by ofborg, even though I am a maintainer of the actual derivation.

I have similar concerns with packages that inherit `meta` from another package, as their `meta.position` will probably be broken

Reply:
The RFC doesn't specify any implementation details. It could be `meta.problems` or some other mechanism. I think we can all agree that it needs to be reliable in the end. This RFCs only sets up a policy what we want to achieve, but not implementation details, because those might change over time and it's hard change existing RFCs to include that.